test/e2e/suite/flake_attempt.go

```diff
@@ -0,0 +1,99 @@
+package suite
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"time"
+
+	"github.com/go-logr/logr"
+	. "github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+	. "github.com/onsi/gomega"
+)
+
+type FlakeAttempt struct {
+	logger logr.Logger
+}
+type FlakeRun struct {
+	DeferCleanup    func(func(context.Context), ...interface{})
+	RetryableExpect func(actual interface{}, extra ...interface{}) Assertion
+}
+
+type IgnorablePanic struct{}
+
+func (IgnorablePanic) GinkgoRecoverShouldIgnoreThisPanic() {}
+
+func (f *FlakeAttempt) It(ctx context.Context, description string, flakeAttempts int, body func(ctx context.Context, flakeRun FlakeRun)) {
+	var lastErr error
+	for attempt := range flakeAttempts {
+		// register globally as well in case the test fails for any reason
+		// including being cancelled while this block is executing
+		// track if the cleanup runs and do not run it again if it does
+		cleanups := []func(context.Context){}
+		deferCleanup := func(cleanup func(context.Context), args ...interface{}) {
+			ran := false
+			onced := func(ctx context.Context) {
+				if ran {
+					f.logger.Info(fmt.Sprintf("Cleanup already ran for flake attempt %d, skipping", attempt+1))
+					return
+				}
+				ran = true
+				cleanup(ctx)
+			}
+			cleanups = append(cleanups, onced)
+			DeferCleanup(onced, args)
+		}
+
+		testGoMega := gomega.NewGomega(func(message string, callerSkip ...int) {
+			lastErr = fmt.Errorf("%s", message)
+			panic(IgnorablePanic{})
+		})
+		flakeRun := FlakeRun{
+			DeferCleanup:    deferCleanup,
+			RetryableExpect: testGoMega.Expect,
+		}
+
+		c := make(chan error)
+		go func() {
+			defer func() {
+				var err error
+				// if this was a ignorable panic, we send nil
+				// to indicate that we should retry
+				// otherwise we send the error to end execution
+				e := recover()
```

Comment:
this is very similar to the ginkgo provided GinkgoRecover. The problem with using it directly is we have no idea, and i could not find a way to determine it from ginkgo global state, if an error has happened and we want to end the test.  This approach allows us to get the error and recover from the panic, but if the error is our retryable error, we "ignore" it.  In the case that the error is not our retryable, then it will fall down into the Eventually and fail the overall test.